server/src/main/java/com/linecorp/centraldogma/server/command/CommandExecutorStatusManager.java

```diff
@@ -0,0 +1,235 @@
+        if (writable != null) {
+            setWritable(writable);
+        }
+        final Boolean replicating = command.replicating();
+        if (replicating != null) {
+            setReplicating(replicating);
+        }
+    }
+
+    /**
+     * Sets the executor to read/write mode or read-only mode.
+     * @return {@code true} if the executor is already in the specified mode, or the mode has been updated
+     *          successfully.
+     */
+    public boolean setWritable(boolean newWritable) {
+        final boolean writable = writable();
+        if (writable == newWritable) {
+            return true;
+        }
+        executor.setWritable(newWritable);
+        if (newWritable) {
+            logger.warn("Left read-only mode.");
+        } else {
+            logger.warn("Entered read-only mode. replication: {}", replicating());
+        }
+        return true;
+    }
+
+    /**
+     * Sets the executor to replicating mode or non-replicating mode.
+     *
+     * <p>This method could take a long time if the executor is not in the desired state yet.
+     * So it should be not called from an event loop thread.
+     *
+     * @return {@code true} if the executor is already in the specified mode, or the mode has been updated
+     *         successfully.
+     */
+    public boolean setReplicating(boolean newReplicating) {
+        if (newReplicating) {
+            try {
+                startExecutor();
+                logger.info("Enabled replication. read-only: {}", !writable());
+                return true;
+            } catch (Exception cause) {
+                logger.warn("Failed to start the command executor:", cause);
+                return false;
+            }
+        } else {
+            try {
+                stopExecutor();
+                logger.info("Disabled replication");
+                return true;
+            } catch (Exception cause) {
+                logger.warn("Failed to stop the command executor:", cause);
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Starts the executor if the {@link CommandExecutor} is not started yet.
+     */
+    private void startExecutor() throws Exception {
+        final CompletableFuture<Void> startFuture;
+        synchronized (this) {
+            switch (replicatingRequest) {
+                case START:
+                    // Start request has already been issued.
+                    return;
+                case STOP:
+                    replicatingRequest = ReplicatingRequest.START;
+                    logger.debug("Cancelled the previous replication stop request");
+                    break;
+                case NONE:
+                    if (replicating()) {
+                        return;
+                    }
+                    replicatingRequest = ReplicatingRequest.START;
+                    break;
+            }
+            startFuture = executor.start();
```

Comment:
Question) What do you think of simply waiting for start/stops inside the `synchronized` block?

```suggestion
            startFuture = executor.start().get();
```

The reasoning is this I think `executor.start()` and `executor.stop()` can be executed concurrently with the current logic.
Looking through `ZooKeeperCommandExecutor` it seems like there are multiple local states, and I'm not sure how thread safe the class is.

Reply:
That is a point of consideration. Initially, as you said, I tried to execute `start` and stop atomically and I didn't want to add this complex state machine.

Migrating the code, I found the following issue that we should call `execute.stop()` to cancel the start process. https://github.com/line/centraldogma/pull/301
We allow the behavior only when a server is being started. Let me try to add a tweak for that.